include/oneapi/dpl/internal/reduce_by_segment_impl.h

```diff
@@ -167,9 +167,9 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
     auto keep_values = __ranges::__get_sycl_range<__bknd::access_mode::read, InputIterator2>();
     auto value_buf = keep_values(first2, first2 + n);
     auto keep_key_outputs = __ranges::__get_sycl_range<__bknd::access_mode::write, OutputIterator1>();
-    auto key_output_buf = keep_key_outputs(result1, result1 + n);
+    auto key_output_buf = keep_key_outputs(result1, dpl::end(result1.get_buffer()));
```

Comment:
result1, result2 can be a USM pointer, host pointer or host iterator....

Reply:
Fixed